lib/rules/one-var.js

```diff
@@ -272,6 +276,87 @@ module.exports = {
             return true;
         }
 
+        /**
+         * Fixer to join VariableDeclaration's into a single declaration
+         * @param   {VariableDeclaration} declarations The `VariableDeclaration` to join
+         * @returns {Function}                         The fixer function
+         */
+        function joinDeclarations(declarations) {
+            const declaration = declarations[0];
+
+            return fixer => {
+                const type = sourceCode.getTokenBefore(declaration);
+                const prevSemi = sourceCode.getTokenBefore(type);
+                const prevType = sourceCode.getTokenBefore(type, {
+                    filter(token) {
+                        return ~["const", "var", "let"].indexOf(token.value) && token.type === "Keyword";
```

Comment:
I don't think this is a reliable way to check the previous statement. For example, with the following code:

```js
var foo = 1;
let bar = function() { var x; };
var baz = 2;
```

`prevType` here will be set to the `var` token *inside* the `bar` function, rather than `let` token that precedes `bar`. As a result, the code will incorrectly get fixed to:

```js
var foo = 1;
let bar = function() { var x; },
 baz = 2;
```

Note that this changes the declaration type of `baz` (it was previously `var` and is now `let`).

To fix this, I would recommend getting the previous statement from the AST rather than scanning through tokens.

Reply:
By "from the AST" do you mean that I should use `getAncestors()`?

Reply:
That would be one solution. Another option would be to use something like `node.parent` to get the parent node in the AST. For example, you could do something like `node.parent[node.parent.body.indexOf(node) - 1]` to get the sibling node that precedes `node`.

I hope this clarifies things, but let me know if it doesn't -- I'm happy to provide more explanation.

Reply:
So I am having trouble getting a node that has node.parent. I run `context.getAncestors(declaration)` which has property `body` when printed out but if I try and do `context.getAncestors(declaration).body` I get undefined. Is there a method like `getTokenBefore` like `getNodeBefore` which gets the previous sibling node of the node that the current token is in or if it is a node then just the previous sibling?

Reply:
I think `context.getAncestors` only returns the ancestors of the node which is currently being traversed in an array, so it doesn't accept an argument.

If you rebase onto the `master` branch, I think all the nodes will have a `parent` property. (There was a [recent change](https://eslint.org/docs/5.0.0/user-guide/migrating-to-5.0.0#parent-before-rules)) that made it easier to use `parent` properties in cases like this.)